File: retrobiocat_web/app/database_bps/curation/routes/ajax/reviewing_ajax.py
import distutils
from distutils.util import strtobool

from flask import request, jsonify
from flask_login import current_user
from flask_security import roles_required

from retrobiocat_web.app.database_bps.curation import bp
from retrobiocat_web.app.database_bps.curation.functions import check_permission
from retrobiocat_web.app.database_bps.curation.routes.ajax.error_results import no_access_to_edit_paper
from retrobiocat_web.app.database_bps.curation.routes.main_tab_routes.overview_tab import get_and_update_status
from retrobiocat_web.mongo.modal_updates import reviewing_updates
from retrobiocat_web.mongo.model_queries import paper_queries
from retrobiocat_web.mongo.models.user_models import user_datastore


@bp.route('/_review_paper_metadata', methods=['GET', 'POST'])
@roles_required('contributor')
def review_paper_metadata():
    """ Mark a paper as reviewed """
    paper = paper_queries.paper_from_id(request.form['paper_id'])
    new_reviewed_status = bool(strtobool(request.form['reviewed']))

    if not check_permission.check_review_persmission(current_user.id, paper):
        return no_access_to_edit_paper()

    issues = reviewing_updates.review_paper_metadata(paper, new_reviewed_status)

    get_and_update_status(paper)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': 'Paper review status updated',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Error while trying to review paper',
                  'issues': issues}

    return jsonify(result=result)

@bp.route('/_sequences_ready_for_review', methods=['GET', 'POST'])
@roles_required('contributor')
def sequences_ready_for_review():
    """Update status of sequences ready to review"""

    paper = paper_queries.paper_from_id(request.form['paper_id'])
    user = user_datastore.get_user(current_user.id)

    if request.form['ready'] == 'true':
        ready = True
    else:
        ready = False

    if check_permission.check_seq_curation_permission(user, paper) == False:
        return no_access_to_edit_paper()

    issues = reviewing_updates.sequences_ready_to_review(paper, ready)

    get_and_update_status(paper)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': f'Paper ready to review sequences status updated (ready={ready})',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Could not update paper ready to review sequences status',
                  'issues': issues}
    return jsonify(result=result)

@bp.route('/_sequences_review', methods=['GET', 'POST'])
@roles_required('contributor')
def sequences_review():
    """Update papers sequences reviewed status"""

    paper = paper_queries.paper_from_id(request.form['paper_id'])
    user = user_datastore.get_user(current_user.id)
    reviewed = bool(distutils.util.strtobool(request.form['reviewed']))

    if reviewed == True:
        if check_permission.check_review_persmission(current_user.id, paper) == False:
            return no_access_to_edit_paper()

    elif reviewed == False:
        if check_permission.has_unreview_access(paper, user) == False:
            result = {'status': 'danger',
                      'msg': 'Can not unreview',
                      'issues': ['Only the owner of the paper can unreview']}
            return jsonify(result=result)

    issues = reviewing_updates.review_paper_sequences(paper, user, reviewed)

    get_and_update_status(paper)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': f'Paper sequence review status updated (reviewed={reviewed})',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Could not update sequence review status',
                  'issues': issues}
    return jsonify(result=result)


@bp.route('/_activity_ready_for_review', methods=['GET', 'POST'])
@roles_required('contributor')
def activity_ready_for_review():
    """Update status of sequences ready to review"""

    paper = paper_queries.paper_from_id(request.form['paper_id'])

    if check_permission.check_paper_permission(current_user.id, paper) == False:
        return no_access_to_edit_paper()

    ready = bool(distutils.util.strtobool(request.form['ready']))
    issues = reviewing_updates.activity_ready_to_review(paper, ready)

    get_and_update_status(paper)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': f'Paper activity data ready to review status updated (ready={ready})',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Could not update paper ready to review activity status',
                  'issues': issues}
    return jsonify(result=result)

@bp.route('/_activity_review', methods=['GET', 'POST'])
@roles_required('contributor')
def activity_review():
    """Update papers sequences reviewed status"""

    paper = paper_queries.paper_from_id(request.form['paper_id'])
    user = user_datastore.get_user(current_user.id)
    reviewed = bool(distutils.util.strtobool(request.form['reviewed']))

    if reviewed == True:
        if check_permission.check_review_persmission(current_user.id, paper) == False:
            return no_access_to_edit_paper()
    elif reviewed == False:
        if check_permission.has_unreview_access(paper, user) == False:
            result = {'status': 'danger',
                      'msg': 'Can not unreview',
                      'issues': ['Only the owner of the paper can unreview']}
            return jsonify(result=result)

    issues = reviewing_updates.review_paper_activity(paper, user, reviewed)

    get_and_update_status(paper)

    if len(issues) == 0:
        result = {'status': 'success',
                  'msg': f'Paper activity review status updated (reviewed={reviewed})',
                  'issues': []}
    else:
        result = {'status': 'danger',
                  'msg': 'Could not update activity review status',
                  'issues': issues}
    return jsonify(result=result)

File: retrobiocat_web/app/database_bps/curation/static/js/ajax/reviewing_ajax.js
var reviewing_ajax = (function () {
    /*
    Methods for making ajax calls for reviewing parts of the curation process
    */


    function review_paper_metadata(paper_id, reviewed, callback, response_div, on_finish) {
        $.post('/_review_paper_metadata', {
            paper_id: paper_id,
            reviewed: reviewed
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                callback()
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying update paper metadata review status', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

    function sequences_ready_review(paper_id, ready, callback, response_div, on_finish) {
        // Mark sequences are ready for review

        $.post('/_sequences_ready_for_review', {
            paper_id: paper_id,
            ready: ready
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                callback(data.result.status)
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying update paper sequences ready review', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

    function sequences_review(paper_id, reviewed, callback, response_div, on_finish) {
        $.post('/_sequences_review', {
            paper_id: paper_id,
            reviewed: reviewed
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                callback(data.result.status)
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying update paper sequences review status', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

        function activity_ready_review(paper_id, ready, callback, response_div, on_finish) {
        $.post('/_activity_ready_for_review', {
            paper_id: paper_id,
            ready: ready
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                callback(data.result.status)
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying update paper activity ready review', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

    function activity_review(paper_id, reviewed, callback, response_div, on_finish) {
        $.post('/_activity_review', {
            paper_id: paper_id,
            reviewed: reviewed
            }).done(function(data) {
                response_msg(data.result.msg, data.result.status, data.result.issues, response_div)
                callback(data.result.status)
            }).fail(function (xhr, status, error) {
                response_msg('JS error while trying update paper activity review status', 'danger', [error], response_div)
                console.log(error)
            }).always(function () {
                if (typeof on_finish !== 'undefined') {
                    on_finish()
                }
            })
        }

    function review_all(paper_id, callback, response_div, on_finish) {
        // run all review routes sequentially

        function paper_callback() {
            paper_sequences_reviewed(paper_id, true, seq_callback, response_div, on_finish)
        }

        function seq_callback(status) {
            paper_activity_reviewed(paper_id, true, callback, response_div, on_finish)
        }

         paper_metadata_update(paper_id, true, paper_callback, response_div, on_finish)
    }


    return {
        review_paper_metadata: review_paper_metadata,
        sequences_ready_review: sequences_ready_review,
        sequences_review: sequences_review,
        activity_ready_review: activity_ready_review,
        activity_review: activity_review,
        review_all: review_all
    }

})()